README.md
<b><i>Unofficial gpSP kai crated by Takka.</b>
--gameplaySP Gameboy Advance emulator for Playstation Portable.</I>

# Release log

-<b>Unofficial gpSP kai 3.4 test 4 build</b>
	Compatible with CFW 5.50GEN-B
	Corrected some sound gaps

-<b>Unofficial gpSP kai 3.4 test 3 build</b>
	Fixed freezing when 1MB of memory cannot be secured when decompressing ZIP (PSP-2000 / 3000)
	Added screen enlargement / reduction settings (only when the main unit is displayed)

-<b>Unofficial gpSP kai 3.4 test 2 build 69</b>
	Adjusted the priority of the main font and sub font (I think the "wo button problem" has been fixed)
	Fixed the GBA folder missing in the distribution file
	Some people cannot install it without preparing the ms0 / PSP / GAME / gpSP folder in the distribution file, so fix it.
	Since the state save data was incompatible, it was corrected so that the old version can also be read.
	Added format version in state save data
	Abolished reading of old state file (506947 byte file)
	Fixed continuous access to Memory Stick during state menu

-<b>Unofficial gpSP kai 3.4 test 1 build 2</b>
	Compatible with ChickHEN R2 (Check 32MB memory usage and ZIP file loading on PSP-2000)
	Fixed a bug when using ZIP files on PSP-1000
	Supports booting from the BIOS (I just want to hear that picone)
	Other minor corrections

-<b>Unofficial gpSP kai 3.3 test 2 build 132 svn.402</b>
	Fixed freeze when loading ZIP file a second time
	Supports reading ZIP files above the ROM buffer
		(Create ZIP_TMP.gba after decompression in the ROM folder. Please be careful about the free space)

-<b>Unofficial gpSP kai 3.3 test 2 build 49 svn.399</b>
	Separate cheats and other options
	Revive C-core ARM emulation for debugging
	Enabled to switch ARM emulation between C core and ASM core (valid after restart)
	Restore the ROM cache that was temporarily reduced to 1MB
	Reduce the RAM cache a little by that amount
	Expand the PSP-1000 decompression buffer to 320KB when decompressing the ZIP file
	BMP buffer on general purpose VRAM, saving about 112.5KB of memory
	Secure 16MB ROM buffer on PSP-1000
	Change icon / image to pochi's
	Added some fonts (included in ZIP file, not network update)
	Fixed freezing when font initialization fails
	Fixed checking other than 0 bit of IME register when interrupting IRQ (it is unknown whether there was a problem due to this)
	Added an option to boot as an old model for debugging (valid after reboot)
	Added an option for debugging (note that the test information will be written to the memory stick)

-<b>Unofficial gpSP kai 3.3 test 1 build 172 svn rev.392</b>
	Fixed R repeated hits not being enabled
	Add build number
	Fixed sound gap during load state
	In case of CFW3.71 M33-3 or less, PSP-2000 does not support TV output / additional memory.
	Supports LF + CR for line feed code in video.cfg
	In order to support CFW3.95GEN, the additional memory is managed by itself as before.
	When decompressing the ZIP file, PSP-2000 uses (maximum) 16MB of decompression buffer (PSP-1000 is 256KB)

-<b>Unofficial gpSP kai 3.2 test 9.0 svn rev.</b>
	Fixed buffer over in message screen title
	Fixed memory corruption when releasing font memory

-<b>Unofficial gpSP kai 3.2 test 8.8 svn rev.</b>
	Fixed running out of memory and hanging when reading a ZIP file on an older model
	Change ZIP file read buffer to VRAM
	Change ZIP file read buffer to 128KB-> 256KB
	Looking at the source of the assembler, it was also optimized by switch, so I returned byte / word writing to the IO register to the switch statement (gcc clever)

-<b>Unofficial gpSP kai 3.2 test 8.7 svn rev.</b>
	The behavior of the "o" and "x" buttons in the menu is now matched to the PSP settings.
	Relaxed cheat name length limit (UTF-8N secures 40 characters)
	Changed memory management in fbm_print.c to malloc / free
	Fixed a hang on exit with the HOME key
	Corrected the byte-by-byte writing process to OAM
	Fixed missing 1 line at 1: 1 output
	Adjust parameters for TV output
	Fixed a video initialization bug
	Deleted the built-in TV output parameters (do not output TV if video.cfg cannot be read normally)
	Fixed some textures wrapping around the edges of the screen (a parameter was avoided until test 8.6)
	Use jump table to write byte / word to IO register to stabilize speed
	Added cheat documentation

-<b>Unofficial gpSP kai 3.2 test 8.6 svn rev.</b>
	Fixed a bug in translucent processing

-<b>Unofficial gpSP kai 3.2 test 8.5 svn rev.</b>
	Fixed a bug in clear_screen ()
	Change the character code of the font file to UCS2
	In line with the above, I converted Naga 10 to UCS2 and converted it to fbm.
	Change the character code of the message file to unicode (UTF-8N)
	Added 10-dot font of Hangul characters
	Change the format of font.cfg
		The first is the main font file
		The second is the sub font file
		Subfont glyphs take precedence
	Removed translucent processing of character display
	Added language setting to "Cheat / Other" options
		It will be enabled at the next startup
	Rename the settings folder to cfg to support updates from the PSP itself
	(It seems that the path name to install is limited to 12 characters)
	Similarly, change the standard installation folder to gpSP
	Similarly, change the GBA save folder to GBA / SAV

	Source code
	Global variables prefixed with g_ (still only part)
	Generalization of scrollbar ()
	Generalization of FILE_READ_MEM and FILE_WRITE_MEM
	Separate the video emulation section and the source around the screen drawing

-<b>Unofficial gpSP kai 3.2 test 8 svn rev.360</b>
	Switching to the menu is a little faster
	Fixed a mistake in switching to the TV screen at startup
	Fixed a keymap read / save error
	Fully adjusted TV output parameters
	A little faster screen_flip ()
	Removed all build settings in kernel mode
	Removed all C-core emulated parts
	Integrate cpu_threaded.c into cpu.c
	Organize savestate related macros
	TV output / ratio 16: 9 / Change settings for enlarged display
	* When outputting TV at ratio 16: 9, black frames are created at the top and bottom, so the resolution is inferior to ratio 4: 3.

-<b>Unofficial gpSP kai 3.2 test 07 svn rev.351</b>
	Change message file
	Added HOME button enable / disable to MENU cheat / other settings
	Added ratio / interlace setting for TV output to graphic / sound settings in MENU
	The screen ratio doesn't do anything internally, just load the corresponding parameters
	Added HOME button setting value to gpsp.cfg
	Added keymap settings to gpsp.cfg
	(The size is 36byte-> 104byte, but it will be updated automatically)
	Updated video.cfg so that the menu screen of the PSP main unit is not filtered
	Keymap can now be switched between default and game specific
	Fixed game cfg not being saved at exit
	Fixed the filter settings not being reflected
	It was not compatible with CFW3.60 of PSP-2000, so it was fixed.

-<b>Unofficial gpSP kai 3.2 test 06 svn rev.346</b>
	Change message file
		Renamed message / font files for Russian
		Added Chinese / Korean message / font file (not translated, content is the same as English)
		Added screen mode option1 / option2 to message file
	Around config
		Added video.cfg
			If the file does not exist, it will start with the standard settings.
			It does not check for errors, so it freezes depending on the parameters.
			As a sample, option 1 is set to 180 degree rotation mode.
		Add header to cfg file
		Changed gpsp config and game config to structs,
		Significant changes around config
		Change key settings, backup settings, and audio buffer settings to game-specific settings
	Around the screen
		Change all parameters around screen settings to structure
		Consolidate and abolish functions around screen settings by the above
		Enabled to expand the menu screen at the time of TV output (current parameters have not been adjusted)
		Subtly added comments around GU
		Double buffered MENU screen to reduce flicker
		GU now clears the screen
		Added 2 types of screen size options (option1, option2)
		video.cfg has screen ratio and interlace settings, but I haven't added it to the MENU, so it's currently 4: 3, non-interlaced.
	Other
		Adjusted GBA memory initialization
		Increased cache buffer (solved some malfunctions, temporary measures, seems to be accessing illegal memory somewhere)
		Fixed the menu / game screen switching at high speed when the HOME key is held down.
		The clock setting value is made into a table

-<b>Unofficial gpSP kai 3.2 test 05 svn rev.319</b>
	Adjusted the stack size to deal with problems when using ZIP files
	Supports menu with HOME button
		(Currently there is no volume display, but it will be implemented in test06)
	Extend and use NJ's homehook.prx so that it can read other than the HOME button
		(The volume button and NOTE button can also be read)
	Significantly changed the directory structure of the source
	I found that prx can use kernel mode specific functions, so I deleted the kernel mode compile setting.
		#Ifdef etc. still remain in the source. Will be deleted in the future.
	Supports full screen of TV output
		Supports PSP-2000 & CFW.71 only
		Regardless of the settings of the main unit, you can switch by connecting the cable and entering and exiting the menu even during the game
		The maximum area of ​​image output seems to be 690x460
		In full screen mode, some of the top, bottom, left, and right are not displayed, but it is a beautiful display because it is an integral multiple.
		In 3: 2 magnified mode, the entire area is displayed
		The internal rendering itself is the same as GBA 240x160
		The operation was confirmed by connecting the D terminal of LCD-TV241XBR-2.
		S / composite has no cable, and the component has no connected device, so I have not confirmed it.
		(I think that it is compatible with the monitor, I would like to read the file as a parameter at the time of TV out in the future)
	Restored the volume of the sound

-<b>Unofficial gpSP kai 3.2 test 04 svn rev.282</b>
	Dealing with freeze bugs when using zip files
	Corresponds to memory corruption during sleep on PSP-2000 (referred to Mr. NJ's comment on HP)

-<b>Unofficial gpSP kai 3.2 test 03 svn rev.278</b>
	Added AHDOC initialization / initialization test
		If WLAN is ON, read, standby test, and disconnect the module.
	Corrected the sound deviation of DirectSound

-<b>Unofficial gpSP kai 3.2 test 02 svn rev.271</b>
	Added firmware version check to determine additional memory
	Removal of thin lines at the top and bottom displayed when magnified
	Split Makefile
	Binary integration into FW3.xx user mode / 64MB version
	Fixed moving to graphics / sound options screen when canceling ROM loading screen

-<b>Unofficial gpSP kai 3.2 test 01 svn rev.260</b>
	Fixed a bug where state load slots were misaligned
	Supports builds for FW3.xx user mode
	Supports increased memory of new PSP
	Build with SDK included with standard PSPSDK + 3.71M33

-<b>Unofficial gpSP kai 3.1 svn rev.252</b>
	Change directory for make
	Fixed the dust display on the right end when enlarged
	Fixed a bug that disturbed the sound
	Fixed a state save / load bug

-<b>Unofficial gpSP kai 3.1 test 18 svn rev.241</b>
	Adjustment at the time of enlarged screen
	Delete GUI alpha blend function / table data
	Added SLOT 10 to state save
	SLOT10 loads / saves to memory without using a memory stick
		The contents will be overwritten when loading / saving to the memory stick.
		Destroyed at exit and when loading roms
	Added CRC32 to ROM INFO
		"0" is displayed for ROMs larger than 16MB
	Added yes / no dialog
	Display dialog when loading / saving state
	The analog ON / OFF setting is also enabled on the menu screen.
	Fixed freeze issue on Mario & Luigi RPG

-<b>Unofficial gpSP kai 3.1 test 17 svn rev.225</b>
	Fixed a slight chipping at the top and bottom when enlarging the screen
	Fixed a bug in save state from the menu screen
	ROM info is always displayed on the menu screen
	Fixed IWRAM reading embug
	Readjust ARM / THUMB ADC / SBC / RSC instructions
		Bug fixes in bit Generations Boundish
	Fixed the screen being distorted in some ROMs
	Fixed noise / screen disturbance during load / save state

-<b>Unofficial gpSP kai 3.1 test 16 svn rev.220</b>
	Fixed the screen being distorted in some ROMs
	Suspend reaction speed improvement
	Supports the display of the progress bar on the GBA screen
	Show progress bar during load / save state
	Add scrollbar

-<b>Unofficial gpSP kai 3.1 test 15 svn rev.213</b>
	Changed to move to the menu when the HOME button is pressed
		Improved response when pressing HOME
	Added display of progress bar when loading
	Supports transparency of palette0 for the time being (supports OBJ where Bit Generations / Dotstream is not displayed)
		There may be a problem with other ROMs, so we are waiting for information.

-<b>Unofficial gpSP kai 3.1 test 14 svn rev.209</b>
	Adjusted frame skip behavior during save state
	UI addition (NJ's CPS2 Emu source has been reused)
		Please forgive that there is no sense in the color scheme. We are waiting for a better color scheme
	If WLAN is ON at startup, perform a connection test
	Fixed HOME and VOL +/- keys not working under heavy load

-<b>Unofficial gpSP kai 3.1 test 13 svn rev.202</b>
	Fixed regular noise
	Fixed ZIP file becoming unreadable

-<b>Unofficial gpSP kai 3.1 test 12 svn rev.198</b>
	Fixed a bug in directory relative references
		Fixed a bug that save data could not be loaded

-<b>Unofficial gpSP kai 3.1 test 11 svn rev.191</b>
	Correspondence of sound noise
	Tilt sensor adjustment
	Supports cheat ROM writing (6aaaaaaa 0000dddd pattern)
		However, only ROM of 16MB or less
		Once enabled, it will not be disabled
	Icon change (provided by frmb0001)
		If you add a background, the title will not be displayed, so leave it out in the test version
	Free translation of BIOS error messages
	Save memory by changing the internal format of message data
	Change the palette to make it a little faster
	Sound thread adjustment
	Adjusted frame skip behavior during load state
	Timer adjustment
	Change directory settings to relative references

-<b>Unofficial gpSP kai 3.1 test 10 svn rev.177</b>
	Sound ON / OFF enabled
	Mounted tilt sensor (can be operated with analog PAD / set all assignments to NONE when using)
	Fine adjustment in various places
	BIOS 0xD / getChecksum implementation

-<b>Unofficial gpSP kai 3.1 test 09 svn rev.170</b>
	Fixed a bug that state save data could not be read
		Sorry, I can't read the data saved in test 8
	Sound thread adjustment

-<b>Unofficial gpSP kai 3.1 test 08 svn rev.165</b>
	Adjusting the sound delay
	Adjusting frame skip processing
	Fixed sound bugs
	Fixed regular noise
	Enabled sound buffer setting

-<b>Unofficial gpSP kai 3.1 test 07 svn rev.150</b>
	Adjusting timer processing
	Adjusting the sound delay
		Sound buffer settings are not currently working, are internal fixation
		We plan to remove the sound buffer settings in the future
		The current setting is twice as accurate as before
		The deviation for each channel is currently under investigation.

-<b>Unofficial gpSP kai 3.1 test 06 svn rev.135</b>
	Fixed a bug when reading a ZIP file

-<b>Unofficial gpSP kai 3.1 test 05.1 svn rev.129</b>
	Fixed sound delay

-<b>Unofficial gpSP kai 3.1 test 05 svn rev.125</b>
	Changed to work in kernel mode
	Made non-SDL around the sound
	Fixed dates display in statesave
	Due to the above, the size of the statesave file has been changed to 506947-> 506951.
		Old files are automatically detected and read when reading
		Will be saved in new format when saved
	Change default directory
	Abolition of split EBOOT.PBP
		If you are using DEVHOOK or FW1.0 ~ 1.5, please do it yourself.
	Many corrections based on the source from the forum thanks ?????

-<b>Unofficial gpSP kai 3.1 test 04 svn rev.65</b>
	Reduction of petit noise
	Changed the number of cheats to 30
	Supports cheat menu page switching
	Fixed cheating not working
	Supports increasing cheats for game cfg files
		Delete the old game cfg file, the new one is 136 bytes

-<b>Unofficial gpSP kai 3.1 test 03 svn rev.45</b>
	Corresponds to freeze at the time of sleeve
		After returning to the sleeve, the menu screen will appear.

-<b>Unofficial gpSP kai 3.1 test 02</b>
	A little optimization
	Returned the sound buffer setting to 2048 ~
	Correspondence of regular petit noise of sound

-<b>Unofficial gpSP kai 3.1 test 01</b>
	Ignore illegal BIOS calls
		Supports some special ROMs
	Adjust a little around the sound
	Readjusted ARM / THUMB SBC / RSC instructions
		Corresponds to the sound of Donkey Kong 3 not coming out

Changes.txt
Please, install "TortoiseSVN" and see produced changes in the source code:
 - https://tortoisesvn.net/downloads.html

src/input.c
/* unofficial gameplaySP kai
 *
 * Copyright (C) 2006 Exophase <exophase@gmail.com>
 * Copyright (C) 2007 takka <takka@tfact.net>
 * Copyright (C) 2007 ????? <?????>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include "common.h"
#include <pspctrl_kernel.h>

// Special thanks to psp298 for the analog->dpad code!

void trigger_key(u32 key);

void trigger_key(u32 key)
{
  u32 p1_cnt = io_registers[REG_P1CNT];
//  u32 p1;

  if((p1_cnt >> 14) & 0x01)
  {
    u32 key_intersection = (p1_cnt & key) & 0x3FF;

    if(p1_cnt >> 15)
    {
      if(key_intersection == (p1_cnt & 0x3FF))
        raise_interrupt(IRQ_KEYPAD);
    }
    else
    {
      if(key_intersection)
        raise_interrupt(IRQ_KEYPAD);
    }
  }
}

u32 key = 0;

extern u32 psp_fps_debug;

#define PSP_ALL_BUTTON_MASK 0x1FFFF

u32 gamepad_config_map[MAX_GAMEPAD_CONFIG_MAP];

u32 last_buttons = 0;
u64 button_repeat_timestamp;

typedef enum
{
  BUTTON_NOT_HELD,
  BUTTON_HELD_INITIAL,
  BUTTON_HELD_REPEAT
} button_repeat_state_type;

button_repeat_state_type button_repeat_state = BUTTON_NOT_HELD;
u32 button_repeat = 0;
gui_action_type cursor_repeat = CURSOR_NONE;

u32 tilt_sensor_x;
u32 tilt_sensor_y;
u32 sensorR;

static u32 button_circle;
static u32 button_cross;

#define BUTTON_REPEAT_START    200000
#define BUTTON_REPEAT_CONTINUE 50000

// GUI用の入力処理
gui_action_type get_gui_input()
{
  SceCtrlData ctrl_data;
  unsigned int ctrl_buttons;
  gui_action_type new_button = CURSOR_NONE;
  u32 new_buttons;
  u32 analog_sensitivity = 92 - (g_gpsp_config.analog_sensitivity_level * 4);
  u32 inv_analog_sensitivity = 256 - analog_sensitivity;

  sceKernelDelayThread(25000);

//  if (quit_flag == 1)
//    quit(0);

  sceCtrlPeekBufferPositive(&ctrl_data, 1);
  if (g_use_home == 1)
  {
    ctrl_buttons = readHomeButton();
    ctrl_data.Buttons |= ctrl_buttons;
  }

  if((g_gpsp_config.enable_analog) && !(ctrl_data.Buttons & PSP_CTRL_HOLD))
  {
    if(ctrl_data.Lx < analog_sensitivity)
      ctrl_data.Buttons = PSP_CTRL_LEFT;

    if(ctrl_data.Lx > inv_analog_sensitivity)
      ctrl_data.Buttons = PSP_CTRL_RIGHT;

    if(ctrl_data.Ly < analog_sensitivity)
      ctrl_data.Buttons = PSP_CTRL_UP;

    if(ctrl_data.Ly > inv_analog_sensitivity)
      ctrl_data.Buttons = PSP_CTRL_DOWN;
  }

  ctrl_data.Buttons &= PSP_ALL_BUTTON_MASK;

  new_buttons = (last_buttons ^ ctrl_data.Buttons) & ctrl_data.Buttons;
  last_buttons = ctrl_data.Buttons;

  if(new_buttons & PSP_CTRL_LEFT)
    new_button = CURSOR_LEFT;

  if(new_buttons & PSP_CTRL_RIGHT)
    new_button = CURSOR_RIGHT;

  if(new_buttons & PSP_CTRL_UP)
    new_button = CURSOR_UP;

  if(new_buttons & PSP_CTRL_DOWN)
    new_button = CURSOR_DOWN;

  if(new_buttons & PSP_CTRL_START)
    new_button = CURSOR_SELECT;

  if(new_buttons & PSP_CTRL_CIRCLE)
    new_button = button_circle;

  if(new_buttons & PSP_CTRL_CROSS)
    new_button = button_cross;

  if(new_buttons & PSP_CTRL_SQUARE)
    new_button = CURSOR_BACK;

  if(new_buttons & PSP_CTRL_RTRIGGER)
    new_button = CURSOR_RTRIGGER;

  if(new_buttons & PSP_CTRL_LTRIGGER)
    new_button = CURSOR_LTRIGGER;

  if(new_buttons & PSP_CTRL_SELECT)
    new_button = KEY_SELECT;

  if(new_buttons & PSP_CTRL_HOME)
    new_button = CURSOR_EXIT;

  if(new_button != CURSOR_NONE)
  {
    sceRtcGetCurrentTick(&button_repeat_timestamp);
    button_repeat_state = BUTTON_HELD_INITIAL;
    button_repeat = new_buttons;
    cursor_repeat = new_button;
  }
  else
  {
    if(ctrl_data.Buttons & button_repeat)
    {
      u64 new_ticks;
      sceRtcGetCurrentTick(&new_ticks);

      if(button_repeat_state == BUTTON_HELD_INITIAL)
      {
        if((new_ticks - button_repeat_timestamp) >
         BUTTON_REPEAT_START)
        {
          new_button = cursor_repeat;
          button_repeat_timestamp = new_ticks;
          button_repeat_state = BUTTON_HELD_REPEAT;
        }
      }

      if(button_repeat_state == BUTTON_HELD_REPEAT)
      {
        if((new_ticks - button_repeat_timestamp) >
         BUTTON_REPEAT_CONTINUE)
        {
          new_button = cursor_repeat;
          button_repeat_timestamp = new_ticks;
        }
      }
    }
  }

  return new_button;
}

#define PSP_CTRL_ANALOG_UP    (1 << 28)
#define PSP_CTRL_ANALOG_DOWN  (1 << 29)
#define PSP_CTRL_ANALOG_LEFT  (1 << 30)
#define PSP_CTRL_ANALOG_RIGHT (1 << 31)

u32 button_psp_mask_to_config[] =
{
  PSP_CTRL_TRIANGLE,
  PSP_CTRL_CROSS,
  PSP_CTRL_CIRCLE,
  PSP_CTRL_SQUARE,
  PSP_CTRL_LTRIGGER,
  PSP_CTRL_RTRIGGER,
  PSP_CTRL_DOWN,
  PSP_CTRL_LEFT,
  PSP_CTRL_UP,
  PSP_CTRL_RIGHT,
  PSP_CTRL_SELECT,
  PSP_CTRL_START,
  PSP_CTRL_ANALOG_UP,
  PSP_CTRL_ANALOG_DOWN,
  PSP_CTRL_ANALOG_LEFT,
  PSP_CTRL_ANALOG_RIGHT
};

u32 button_id_to_gba_mask[] =
{
  BUTTON_UP,
  BUTTON_DOWN,
  BUTTON_LEFT,
  BUTTON_RIGHT,
  BUTTON_A,
  BUTTON_B,
  BUTTON_L,
  BUTTON_R,
  BUTTON_START,
  BUTTON_SELECT,
  BUTTON_NONE,
  BUTTON_NONE,
  BUTTON_NONE,
  BUTTON_NONE
};

gui_action_type get_gui_input_fs_hold(u32 button_id)
{
  gui_action_type new_button = get_gui_input();
  if((last_buttons & button_psp_mask_to_config[button_id]) == 0)
    return CURSOR_BACK;

  return new_button;
}

u32 rapidfire_flag = 1;

// エミュレーション時の入力処理
u32 update_input()
{
  SceCtrlData ctrl_data;
  u32 buttons;
  u32 ctrl_buttons = 0;
  u32 non_repeat_buttons;
  u32 button_id = 0;
  u32 i;
  u32 new_key = 0;
  u32 analog_sensitivity = 92 - (g_gpsp_config.analog_sensitivity_level * 4);
  u32 inv_analog_sensitivity = 256 - analog_sensitivity;
//  tilt_sensor_x = 0x800;
//  tilt_sensor_y = 0x800;
//  sensorR = 0x650;

  sceCtrlPeekBufferPositive(&ctrl_data, 1);

  if (g_use_home == 1)
    ctrl_buttons = readHomeButton();

  buttons = ctrl_data.Buttons | ctrl_buttons;

  if((g_gpsp_config.enable_analog) && !(buttons & PSP_CTRL_HOLD))
  {
    if(g_game_config.allocate_sensor == NO)
    {
      if(ctrl_data.Lx < analog_sensitivity)
        buttons |= PSP_CTRL_ANALOG_LEFT;

      if(ctrl_data.Lx > inv_analog_sensitivity)
        buttons |= PSP_CTRL_ANALOG_RIGHT;

      if(ctrl_data.Ly < analog_sensitivity)
        buttons |= PSP_CTRL_ANALOG_UP;

      if(ctrl_data.Ly > inv_analog_sensitivity)
        buttons |= PSP_CTRL_ANALOG_DOWN;
    }
    else
    {
      tilt_sensor_x = ctrl_data.Lx * 16;    // センター 2048(0x800) 最小値 0(0x0) 最大値 1143(0xFFF) 幅 4096
      tilt_sensor_y = ctrl_data.Ly * 16;    // センター 2048(0x800) 最小値 0(0x0) 最大値 1152(0xFFF) 幅 4096
      //    sensorR = ctrl_data.Lx * 12.5;  // センター 1600(0x640) 最小値 0(0x0) 最大値 3200(0xC80) 幅 3200
    }
  }

  non_repeat_buttons = (last_buttons ^ buttons) & buttons;
  last_buttons = buttons;

  for(i = 0; i < 16; i++)
  {
    if(non_repeat_buttons & button_psp_mask_to_config[i])
      button_id = gamepad_config_map[i];
    // HOMEが押されたらMENUに移行
    if (buttons & PSP_CTRL_HOME) button_id = BUTTON_ID_MENU;

    switch(button_id)
    {
      case BUTTON_ID_MENU:
      {
        u16 *screen_copy = copy_screen();
        u32 ret_val = menu(screen_copy);
        free(screen_copy);

        return ret_val;
      }

      case BUTTON_ID_LOADSTATE:
      {
        char current_savestate_filename[MAX_FILE];
        get_savestate_filename_noshot(g_savestate_slot_num, current_savestate_filename);
        if (load_state(current_savestate_filename, g_savestate_slot_num) == 1)
          return 1;
        else
        {
          pause_sound(0);
          return 0;
        }
      }

      case BUTTON_ID_SAVESTATE:
      {
        char current_savestate_filename[MAX_FILE];
        u16 *current_screen = copy_screen();
        u32 ret;
        get_savestate_filename_noshot(g_savestate_slot_num, current_savestate_filename);
        ret = save_state(current_savestate_filename, current_screen, g_savestate_slot_num);
        free(current_screen);
        if (ret == 1)
          return 0;
        else
        {
          pause_sound(0);
          return 0;
        }
      }

      case BUTTON_ID_FASTFORWARD:
//        PRINT_STRING_BG("FASTFORWARD", 0xFFFF, 0x0000, 0, 50);
        g_synchronize_flag ^= 1;
        break;
//        return 0;

      case BUTTON_ID_VOLUP:
//        gp2x_sound_volume(1);
        break;

      case BUTTON_ID_VOLDOWN:
//        gp2x_sound_volume(0);
        break;

      case BUTTON_ID_FPS:
        psp_fps_debug ^= 1;
        break;
    }

    if(buttons & button_psp_mask_to_config[i])
    {
      button_id = gamepad_config_map[i];
      if(button_id < BUTTON_ID_MENU)
      {
        new_key |= button_id_to_gba_mask[button_id];
      }
      else

      if((button_id >= BUTTON_ID_RAPIDFIRE_A) &&
       (button_id <= BUTTON_ID_RAPIDFIRE_R))
      {
        rapidfire_flag ^= 1;
        if(rapidfire_flag)
        {
          new_key |= button_id_to_gba_mask[button_id - BUTTON_ID_RAPIDFIRE_A + BUTTON_ID_A];
        }
        else
        {
          new_key &= ~button_id_to_gba_mask[button_id - BUTTON_ID_RAPIDFIRE_A + BUTTON_ID_A];
        }
      }
    }
  }

  if((new_key | key) != key)
    trigger_key(new_key);

  key = new_key;

  io_registers[REG_P1] = (~key) & 0x3FF;

  return 0;
}

void init_input()
{
  int button_swap;
  sceCtrlSetSamplingCycle(0);
  sceCtrlSetSamplingMode(PSP_CTRL_MODE_ANALOG);
// TODO:GEN-B
  if (g_use_home == 1)
    initHomeButton(sceKernelDevkitVersion());
  tilt_sensor_x = 0x800;
  tilt_sensor_y = 0x800;
  sceUtilityGetSystemParamInt(9, &button_swap);
  if (button_swap == 0)
  {
    button_circle = CURSOR_SELECT;
    button_cross  = CURSOR_EXIT;
  }
  else
  {
    button_circle = CURSOR_EXIT;
    button_cross  = CURSOR_SELECT;
  }
}

// type = READ / WRITE_MEM
#define input_savestate_body(type, ver)                                       \
{                                                                             \
  FILE_##type##_VARIABLE(g_state_buffer_ptr, key);                            \
}                                                                             \

void input_read_mem_savestate(u32 ver)
  input_savestate_body(READ_MEM, ver);

void input_write_mem_savestate(u32 ver)
  input_savestate_body(WRITE_MEM, ver);

// 以降OSK用のコード
#ifdef OSK

static unsigned int __attribute__((aligned(16))) list[262144];

#define BUF_WIDTH (512)
#define SCR_WIDTH (480)
#define SCR_HEIGHT (272)
#define PIXEL_SIZE (4) /* change this if you change to another screenmode */
#define FRAME_SIZE (BUF_WIDTH * SCR_HEIGHT * PIXEL_SIZE)
#define ZBUF_SIZE (BUF_WIDTH SCR_HEIGHT * 2) /* zbuffer seems to be 16-bit? */

int SetupCallbacks();
void SetupGu();

int main(int argc, char* argv[])
{
     int done = 0;
     SetupCallbacks();

     // INIT GU!!!! it will not work without!!!
     SetupGu();

     // INIT OSK
     unsigned short intext[128]  = { 0 }; // text already in the edit box on start
     unsigned short outtext[128] = { 0 }; // text after input
     unsigned short desc[128]    = { 'E', 'n', 't', 'e', 'r', ' ', 'T', 'e', 'x', 't', 0 }; // description

     SceUtilityOskData data;
     memset(&data, 0, sizeof(data));
     data.language = 2; // english
     data.lines = 1; // just online
     data.unk_24 = 1; // set to 1
     data.desc = desc;
     data.intext = intext;
     data.outtextlength = 128; // sizeof(outtext) / sizeof(unsigned short)
     data.outtextlimit = 32; // just allow 32 chars
     data.outtext = outtext;

     SceUtilityOskParams osk;
     memset(&osk, 0, sizeof(osk));
     osk.size = sizeof(osk);
     osk.language = 2;
     osk.buttonswap = 0;
     osk.unk_12 = 17; // What
     osk.unk_16 = 19; // the
     osk.unk_20 = 18; // fuck
     osk.unk_24 = 16; // ???
     osk.unk_48 = 1;
     osk.data = &data;

     // Only ascii code is handled so only the input of the small letters is printed

     if(sceUtilityOskInitStart(&osk))
     {
          return 0;
     }

     while(!done)
     {
          sceGuStart(GU_DIRECT,list);
          sceGuClearColor(0x666666);
          sceGuClearDepth(0);
          sceGuClear(GU_COLOR_BUFFER_BIT | GU_DEPTH_BUFFER_BIT);

          sceGuFinish();
          sceGuSync(0,0);

          switch(sceUtilityOskGetStatus())
          {
          case PSP_OSK_NONE:
               break;
          case PSP_OSK_INIT:
               break;
          case PSP_OSK_VISIBLE:
               sceUtilityOskUpdate(2); // 2 is taken from ps2dev.org recommendation
               break;
          case PSP_OSK_QUIT:
               sceUtilityOskShutdownStart();
               break;
          case PSP_OSK_FINISHED:
               done = 1;
               break;
          default:
               break;
          }

          sceDisplayWaitVblankStart();
          sceGuSwapBuffers();
     }

     pspDebugScreenInit();
     pspDebugScreenSetXY(0, 0);
     pspDebugScreenPrintf("S:");
     int i;
     for(i = 0; data.outtext[i]; i++)
     {
          unsigned c = data.outtext[i];
          if(32 <= c && c <= 127)
          {
               pspDebugScreenPrintf("%c", data.outtext[i]); // print ascii only
          }
     }
     pspDebugScreenPrintf("\n");

     sceKernelSleepThread();
     return 0;
}

void SetupGu()
{
     sceGuInit();

     sceGuStart(GU_DIRECT,list);
     sceGuDrawBuffer(GU_PSM_8888,(void*)0,BUF_WIDTH);
     sceGuDispBuffer(SCR_WIDTH,SCR_HEIGHT,(void*)0x88000,BUF_WIDTH);
     sceGuDepthBuffer((void*)0x110000,BUF_WIDTH);
     sceGuOffset(2048 - (SCR_WIDTH/2),2048 - (SCR_HEIGHT/2));
     sceGuViewport(2048,2048,SCR_WIDTH,SCR_HEIGHT);
     sceGuDepthRange(0xc350,0x2710);
     sceGuScissor(0,0,SCR_WIDTH,SCR_HEIGHT);
     sceGuEnable(GU_SCISSOR_TEST);
     sceGuDepthFunc(GU_GEQUAL);
     sceGuEnable(GU_DEPTH_TEST);
     sceGuFrontFace(GU_CW);
     sceGuShadeModel(GU_FLAT);
     sceGuEnable(GU_CULL_FACE);
     sceGuEnable(GU_TEXTURE_2D);
     sceGuEnable(GU_CLIP_PLANES);
     sceGuFinish();
     sceGuSync(0,0);

     sceDisplayWaitVblankStart();
     sceGuDisplay(GU_TRUE);
}

int exit_callback(int arg1, int arg2, void *common)
{
     sceGuTerm();
     sceKernelExitGame();
     return 0;
}

int CallbackThread(SceSize args, void *argp)
{
     int cbid = sceKernelCreateCallback("Exit Callback", exit_callback, NULL);
     sceKernelRegisterExitCallback(cbid);
     sceKernelSleepThreadCB();
     return 0;
}

int SetupCallbacks(void)
{
     int thid = sceKernelCreateThread("update_thread", CallbackThread, 0x11, 0xFA0, 0, 0);
     if(thid >= 0) {
          sceKernelStartThread(thid, 0, 0);
     }
     return thid;
}
#endif
